spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Graph.scala

```diff
@@ -175,26 +186,110 @@ class Graph[T: ClassTag](val inputs : Seq[ModuleNode[T]],
       "A graph container should not be changed after it is constructed")
   }
 
+  // input node for backward graph
+  val backInputs = new Array[ModuleNode[T]](inputs.length)
+
   // Add a dummy output node, to get an one end graph. So the nodes that are not dependent by
   // the outputs will be excluded
   private val dummyOutput = new ModuleNode[T](new Dummy[T]())
+  // dummy output of backward graph
+  private var dummyOutputGrad: ModuleNode[T] = null
   outputs.foreach(_ -> dummyOutput)
 
   /**
    * Computing backgraph
    */
   val backGraph = dummyOutput.graph(reverse = true)
+  var gradGraph: DirectedGraph[AbstractModule[Activity, Tensor[T], T]] = null
 
   /**
    * Execution plan
    */
   val executions = backGraph.topologySort.filter(!_.element.isInstanceOf[Dummy[T]]).reverse
+  var executionsGrad: Array[Node[AbstractModule[Activity, Tensor[T], T]]] = null
+
   modules.appendAll(executions.map(_.element.asInstanceOf[AbstractModule[Activity, Activity, T]]))
 
+
+
+  private def setTrainable(): Unit = {
+    var i = executions.length - 1
+    while (i >= 0) {
+      val node = executions(i)
+      if (node.nextNodes.length > 1) {
+        node.element.trainable = node.element.trainable &&
+          node.nextNodes.exists(_.element.trainable)
+      }
+      if (!node.element.trainable) {
+        node.prevNodes.foreach(prev => {
+          if (prev.nextNodes.length == 1) {
+            prev.element.trainable = false
+          }
+        })
+      }
+      i -= 1
+    }
+    i = 0
+    while (i < executions.length) {
+      val node = executions(i)
+      if (!needBackward(node)) {
+        node.element.trainable = false
+      }
+      if (!node.element.trainable) {
+        node.nextNodes.foreach(next => {
+          if (!needBackward(next)) next.element.trainable = false
+        })
+      }
+      i += 1
+    }
+  }
+
+  private def needBackward(node: ModuleNode[T]): Boolean = {
+    if (!node.element.trainable) return false
+    if (node.element.parameters() != null) return true
+    node.prevNodes.foreach(n => {
+      if (needBackward(n)) return true
+    })
+    false
+  }
+
+  def build(): this.type = {
+    setTrainable()
```

Comment:
mark the non-trainable nodes from input node to output node
 For example, Reshape() -> Identity() -> Conv() -> SoftMax()
Here the Reshape and Identity can be non-trainable, since they do not need to propagate the
gradients to any layers with parameters